### PR TITLE
Switch output dependency from TAP to Teamcity

### PR DESCRIPTION
**Pull request by padraic**
Just some simple changes. I disabled early termination of the ProcessRunner since it shouldn't be needed if we are telling PHPUnit to stop on first failure (via command line option), and it ends up truncating the failure messages when we do it manually.

However, I think the `\Humbug\ProcessRunner` needs a bit of attention beyond this. I've commented out lines, and there seems to some interference when underlying unit-under-test is also using processes. You can spot it by attempting to use Humbug on https://github.com/Seldaek/monolog (for example). The failure between Humbug, and running PHPUnit straight on Monolog, are very different, but you can switch failures by manipulating the ProcessRunner. This was happening WITHOUT my changes, so it's a separate pre-existing issue.

Anyways, will add tests and such over next day or so.

### CODE REVIEW

**Comment by padraic** [comment]
Tests have a single failure which seems to have crept in via a possible PHPUnit bug from PHPUnit 5.2.0 onwards. I'm not sure if it's expected or not, but I've reported it here: https://github.com/sebastianbergmann/phpunit/issues/2665

If the outcome is something expected/deprecated or otherwise not to be fixed, I'll scrub the offending test.

**Comment by omissis** [comment]
@padraic I was wondering if it might make sense to have adapters for each logging format, so to avoid couple the implementation to a specific one and be able to swap between them more easily.

**Comment by padraic** [comment]
@omissis That makes sense. There was a reason originally for the tight coupling, but that's no longer an issue.

**Comment by padraic** [comment]
@humbug/core Ready for review here now. There is one test now marked skipped - against the old PHPUnit suite assembly method using AllTests.php files. Can't quite find it documented in the PHPUnit changelog, but Sebastian certainly doesn't consider this a supported feature anymore.

File: tests/Adapter/_files/regression/server-argv/phpunit.xml.dist [hunk 1]
```diff
@@ -17,5 +17,11 @@
...
     </testsuites>
-    
+    <filter>
+        <whitelist>
+            <exclude>
+                <directory>.</directory>
+            </exclude>
+        </whitelist>
+    </filter>
 </phpunit>
```
Comment (theofidry) [inline tests/Adapter/_files/regression/server-argv/phpunit.xml.dist]: missing ending blank line

File: features/running/developer_sees_initial_testrun_error_if_suite_not_passing.feature [hunk 2]
```diff
@@ -68,13 +68,5 @@ Feature: Use Humbug
...
-            Humbug running test suite to generate logs and code coverage data...
-
-            Tests must be in a fully passing state before Humbug is run.
-            Incomplete, skipped or risky tests are allowed.
-            The testing framework reported an exit code of 143.
-            The testing framework ran into a failure or error. Refer to output below.
-            Stdout:
-               > TAP version 13
-               > not ok 1 - Error: FooTest::testAddsNumbers
+            ##teamcity[testFailed
```
Comment (theofidry) [inline features/running/developer_sees_initial_testrun_error_if_suite_not_passing.feature]: we no longer get to know where it's failing?

File: src/Renderer/Text.php [hunk 3]
```diff
@@ -316,4 +316,13 @@ protected function headAndTail($output, $lineCount = 20, $omittedMarker = '[...M
             array_slice($lines, -$lineCount, $lineCount)
         ));
     }
+
+    protected function extractFail($output)
+    {
+        $result = preg_match('%##teamcity\[testFailed.*\]%', $output, $matches);
```
Comment (theofidry) [inline src/Renderer/Text.php]:
```php
if (preg_match('%##teamcity\[testFailed.*\]%', $output, $matches)) {
    return $matches[0];
}

return 'No failure output was detected by Humbug, but a failure was reported by PHPUnit.';
```

Comment (theofidry) [inline src/Renderer/Text.php]: Could also be simplified by a ternary operator

File: src/Renderer/Text.php [hunk 4]
```diff
@@ -316,4 +316,13 @@ protected function headAndTail($output, $lineCount = 20, $omittedMarker = '[...M
             array_slice($lines, -$lineCount, $lineCount)
         ));
     }
+
+    protected function extractFail($output)
```
Comment (theofidry) [inline src/Renderer/Text.php]: can we make it private?

File: tests/Adapter/_files/phpunit.xml.dist [hunk 5]
```diff
@@ -22,7 +22,9 @@
 
     <filter>
         <whitelist>
-            <directory>.</directory>
+            <exclude>
```
Comment (theofidry) [inline tests/Adapter/_files/phpunit.xml.dist]: hm why that change?

File: src/Adapter/AdapterAbstract.php [hunk 6]
```diff
@@ -52,12 +54,8 @@
      */
     public function ok($output)
     {
-        $lines = explode("\n", $output);
-        foreach ($lines as $line) {
-            if (preg_match("%not ok \\d+ - %", $line)
-            && !preg_match("%# TODO%", $line)) {
-                return false;
-            }
+        if (preg_match("%##teamcity\[testFailed%", $output)) {
```
Comment (theofidry) [inline src/Adapter/AdapterAbstract.php]:
```php
return (bool) preg_match("%##teamcity\[testFailed%", $output);
```

File: tests/Adapter/PhpunitTest.php [hunk 7]
```diff
@@ -88,12 +88,13 @@ public function testAdapterRunsDefaultPhpunitCommand()
...
         $result = $process->getOutput();
 
-        $this->assertStringStartsWith('TAP version', $result, $process->getErrorOutput());
+        $this->assertContains('##teamcity[', $result, $process->getErrorOutput());
         $this->assertTrue($adapter->ok($result));
     }
 
     public function testAdapterRunsPhpunitCommandWithAlltestsFileTarget()
     {
+        $this->markTestSkipped('See https://github.com/sebastianbergmann/phpunit/issues/2665');
```
Comment (theofidry) [inline tests/Adapter/PhpunitTest.php]: looks like this issue has been closed and that we shouldn't run that test

File: features/running/developer_sees_initial_testrun_error_if_suite_not_passing.feature (hunk 2 above)
Comment (padraic) [inline features/running/developer_sees_initial_testrun_error_if_suite_not_passing.feature]: The teamcity block will have it - but the block varies between test runs, so I just checked that it was output for the scenario..

File: tests/Adapter/PhpunitTest.php (hunk 7 above)
Comment (padraic) [inline tests/Adapter/PhpunitTest.php]: True - test should actually be deleted now.

File: src/Adapter/AdapterAbstract.php [hunk 8]
```diff
@@ -67,16 +65,15 @@ public function ok($output)
...
      * @return bool|int
      */
     public function hasOks($output)
     {
-        $result = preg_match_all("%ok (\\d+) - .*%m", $output, $matches);
+        $result = preg_match_all("%##teamcity\[testFinished%", $output);
         if ($result) {
-            return (int) end($matches[1]);
+            $this->okCount += $result;
+            return $this->okCount; // was: (int) end($matches[1]);
```
Comment (omissis) [inline src/Adapter/AdapterAbstract.php]: I would expect a has* method to always return a boolean

File: src/Adapter/AdapterAbstract.php (hunk 6 above)
Comment (padraic) [inline src/Adapter/AdapterAbstract.php]: You have to watch the casting to bool with preg_match - that's why I prefer taking the scenic route ;).

**Comment by padraic** [comment]
@theofidry Some changes made.

File: src/Adapter/AdapterAbstract.php (hunk 8 above)
Comment (padraic) [inline src/Adapter/AdapterAbstract.php]: It probably should, and it's a public method to boot, but something we can clean up with greater forethought for 2.0 when defining interfaces/extension points.

File: src/Adapter/AdapterAbstract.php (hunk 6 above)
Comment (theofidry) [inline src/Adapter/AdapterAbstract.php]: but isn't it what is done with the if statement? Too me both look strictly equivalent

Comment (padraic) [inline src/Adapter/AdapterAbstract.php]: Here's the short version that would work (excl. errors which return FALSE):
```php
return 0 === preg_match("%##teamcity\[testFailed%", $output);
```
Added via commit just now.

**Comment by padraic** [comment]
@theofidry In case my comment was lost in changing that, I updated the preg_match block to
```php
return 0 === preg_match("%##teamcity\[testFailed%", $output);
```

**Pull request merged**